src/Ledger/Utxo.lagda

```diff
@@ -40,15 +40,25 @@ infixl 7 _*↓_
 _*↓_ : ℚ.ℚ → ℕ → ℕ
 q *↓ n = ℤ.∣ ℚ.⌊ q ℚ.* (ℤ.+ n ℚ./ 1) ⌋ ∣
 
-isPhaseTwoScriptAddress : Tx → Addr → Bool
-isPhaseTwoScriptAddress tx a =
+isTwoPhaseScriptAddress : Tx → UTxO → Addr → Bool
+isTwoPhaseScriptAddress tx utxo a =
   if isScriptAddr a then
-    (λ {p} → if lookupScriptHash (getScriptHash a p) tx
-                 then (λ {s} → isP2Script s)
-                 else false)
+      (λ {p} → if lookupScriptHash (getScriptHash a p) tx utxo then
+               (λ {s} → isP2Script s)
+             else false)
   else
     false
 
+opaque
+  getDataHashes : ℙ TxOut → ℙ DataHash
+  getDataHashes txo = mapPartial isInj₂ (mapPartial (proj₁ ∘ proj₂ ∘ proj₂) txo)
+
+  getInputHashes : Tx → UTxO → ℙ DataHash
+  getInputHashes tx utxo = getDataHashes
+    (filterˢ (λ (a , _ ) → isTwoPhaseScriptAddress tx utxo a ≡ true)
+            (range (utxo ∣ txins)))
+    where open Tx; open TxBody (tx .body)
+
```

Comment:
This block should be in a figure, and I think while you're at it you could move `totExUnits` in there as well. Also, make sure that the `opaque` keyword is hidden from the figure.